src/charm.py

```diff
@@ -463,10 +467,56 @@ def restart_grafana(self) -> None:
                 "not exist or is not responsive"
             )
 
+    def _parse_grafana_path(self, parts: ParseResult) -> dict:
+        """Convert web_external_url into a usable path."""
+        ret = {}
+
+        # urlparse.path parsing is absolutely horrid and only
+        # guarantees any kind of sanity if there is a scheme
+        if parts.scheme:
+            ret["scheme"] = parts.scheme
+            ret["host"] = "0.0.0.0"
+            ret["port"] = parts.netloc.split(":")[1] if ":" in parts.netloc else "3000"
+            ret["path"] = parts.path
+        elif parts.path.startswith("/"):
+            ret["scheme"] = "http"
+            ret["host"] = "0.0.0.0"
+            ret["port"] = "3000"
+            ret["path"] = parts.path
+        else:
+            # This could really be anything!
+            logger.warning(
+                "Could not determine web_external_url for Grafana. Please "
+                "use a fully-qualified path or a bare subpath"
+            )
+
```

Comment:
```suggestion
        if not parts.scheme and not parts.path.startswith("/"):
            # This could really be anything!
            logger.warning(
                "Could not determine web_external_url for Grafana. Please "
                "use a fully-qualified path or a bare subpath"
            )
            return {}
            
        return {
            "scheme": parts.scheme or "http",
            "host": "0.0.0.0",
            "port": parts.netloc.split(":")[1] if ":" in parts.netloc else "3000",
            "path": parts.path,
        }
```

This should be fairly equivalent, but maybe a bit easier to reason about. Feel free to just reject it if you prefer what you already have.

Reply:
I like this --- admittedly, I'm more or less waiting for confirmation of "does this work"?

I should setup a microk8s instance with an ingress and test it, but the patch is up to unblock internal testing, and it's that useful to any consumers until we resolve the ingress story